File: week12/추억 점수.js
function solution(name, yearning, photo) {
  const yearningScoreMap = {};
  name.forEach((person, i) => {
    yearningScoreMap[person] = yearning[i];
  });

  const answer = photo.map(persons => persons.reduce((sum, person) => sum + (yearningScoreMap[person] || 0), 0));

  return answer;
}


function solution(name, yearning, photo) {
    return photo.map(v=> v.reduce((sum, current)=> sum += yearning[name.indexOf(current)] ?? 0, 0))
}


console.log(solution(
  ["may", "kein", "kain", "radi"],
  [5, 10, 1, 3],
  [["may", "kein", "kain", "radi"], ["may", "kein", "brin", "deny"], ["kon", "kain", "may", "coni"]]
)); // [19, 15, 6]

console.log(solution(
  ["kali", "mari", "don"],
  [11, 1, 55],
  [["kali", "mari", "don"], ["pony", "tom", "teddy"], ["con", "mona", "don"]]
)); // [67, 0, 55]

console.log(solution(
  ["may", "kein", "kain", "radi"],
  [5, 10, 1, 3],
  [["may"], ["kein", "deny", "may"], ["kon", "coni"]]
)); // [5, 15, 0]

File: 추억점수.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> solution(vector<string> name, vector<int> yearning, vector<vector<string>> photo)
{
    vector<int> ans; ans.reserve(photo.size());
    unordered_map<string, int> score; score.reserve(name.size());
    
    int i = 0;
    for (const auto& n : name)
        score[n] = yearning[i++];
    
    int j = 0;
    for (const auto& p : photo)
    {
        ans.push_back(0);
        for (const auto& n : p)
            if (score.find(n) != score.end())
                ans[j] += score[n];
        j++;
    }
    
    return ans;
}
